src/apps/js_generic/js_generic.cpp

```diff
@@ -436,6 +436,54 @@ namespace ccfapp
     return JS_ThrowTypeError(ctx, "Cannot call set on read-only map");
   }
 
+  static JSValue js_kv_map_foreach(
+    JSContext* ctx, JSValueConst this_val, int argc, JSValueConst* argv)
+  {
+    auto map_view =
+      static_cast<KVMap::TxView*>(JS_GetOpaque(this_val, kv_map_view_class_id));
+
+    if (argc != 1)
+      return JS_ThrowTypeError(
+        ctx, "Passed %d arguments, but expected 1", argc);
+
+    JSValue func = argv[0];
+
+    if (!JS_IsFunction(ctx, func))
+      return JS_ThrowTypeError(ctx, "Argument must be a function");
+
+    bool failed = false;
+    map_view->foreach([ctx, func, &failed](const auto& k, const auto& v) {
+      JSValue args[2];
+      args[0] = JS_NewArrayBufferCopy(ctx, k.data(), k.size());
+      args[1] = JS_NewArrayBufferCopy(ctx, v.data(), v.size());
+
+      auto val = JS_Call(ctx, func, JS_UNDEFINED, 2, args);
+
+      JS_FreeValue(ctx, args[0]);
+      JS_FreeValue(ctx, args[1]);
+
+      if (JS_IsException(val))
+      {
+        js_dump_error(ctx);
+        failed = true;
+        return false;
+      }
+
+      JS_FreeValue(ctx, val);
+
+      return true;
+    });
+
+    JS_FreeValue(ctx, func);
+
+    if (failed)
+    {
+      return JS_ThrowInternalError(ctx, "foreach loop failed");
```

Comment:
I don't think you need to throw another error since the one from the loop is still current. Probably enough to `return JS_Exception` here.

Reply:
So I thought it should be possible to `return JS_GetException(ctx);` here, but it's not. First we're only allowed to call `GetException` once, and we already do so in `dump_error`. Even if we fiddle to call this only once, it looks like returning this `JSValue` from this function doesn't work, causing errors when QuickJS later tries to free things. It's happy with `JS_ThrowXXX` values, or `JS_EXCEPTION`, but not `JS_GetException`, so sticking with the current result.

Reply:
What's the issue with `JS_EXCEPTION`? This is what `JS_ThrowInternalError` (or any other exception result) returns anyway.

Reply:
You're right, this `JS_EXCEPTION` const works fine.